rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,197 @@
+import org.thingsboard.rule.engine.api.TbNode;
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.rule.engine.api.util.TbNodeUtils;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.id.DeviceId;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.msg.TbMsgType;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+import org.thingsboard.server.common.msg.queue.PartitionChangeMsg;
+import org.thingsboard.server.common.msg.queue.TbCallback;
+
+import java.time.Duration;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.ACTION,
+        name = "device state",
+        nodeDescription = "Triggers device connectivity events",
+        nodeDetails = "If incoming message originator is a device, registers configured event for that device in the Device State Service, which sends appropriate message to the Rule Engine." +
+                " If metadata <code>ts</code> property is present, it will be used as event timestamp. Otherwise, the message timestamp will be used." +
+                " Incoming message is forwarded using the <code>Success</code> chain, unless an unexpected error occurs during message processing" +
+                " then incoming message is forwarded using the <code>Failure</code> chain." +
+                "<br>" +
+                "Supported device connectivity events are:" +
+                "<ul>" +
+                "<li>Connect event</li>" +
+                "<li>Disconnect event</li>" +
+                "<li>Activity event</li>" +
+                "<li>Inactivity event</li>" +
+                "</ul>" +
+                "This node is particularly useful when device isn't using transports to receive data, such as when fetching data from external API or computing new data within the rule chain.",
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    private static final Set<TbMsgType> SUPPORTED_EVENTS = EnumSet.of(
+            TbMsgType.CONNECT_EVENT, TbMsgType.ACTIVITY_EVENT, TbMsgType.DISCONNECT_EVENT, TbMsgType.INACTIVITY_EVENT
+    );
+    private static final Duration ONE_SECOND = Duration.ofSeconds(1L);
+    private static final Duration ENTRY_EXPIRATION_TIME = Duration.ofDays(1L);
+    private static final Duration ENTRY_CLEANUP_PERIOD = Duration.ofHours(1L);
+
+    private Stopwatch stopwatch;
+    private ConcurrentMap<DeviceId, Duration> lastActivityEventTimestamps;
```

Comment:
ConcurrentReferenceHashMap<DeviceId, TbRateLimits>  with SOFT
Make the rate limit configuration an env parameter. Default value: 1:1,30:60,60:3600.